src/OmniSharp/Workers/Refactoring/FixUsings.cs

```diff
@@ -0,0 +1,238 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using OmniSharp.Models;
+using OmniSharp.Services;
+
+namespace OmniSharp
+{
+    public class FixUsingsWorker
+    {
+        public static async Task<FixUsingsResponse> FixUsings(OmnisharpWorkspace workspace, string fileName, Document document, SemanticModel semanticModel)
+        {
+            var codeActionProvider = new RoslynCodeActionProvider();
+            var ambiguous = await AddMissingUsings(fileName, workspace, document, semanticModel, codeActionProvider);
+            await RemoveDuplicateUsings(fileName, workspace, document, semanticModel, codeActionProvider);
+            document = workspace.GetDocument(fileName);
+            var compilationUnitSyntax = (await AddLinqQuerySyntax(fileName, workspace, document, semanticModel));
+            document = document.WithSyntaxRoot(compilationUnitSyntax);
+            var docText = await document.GetTextAsync();
+
+            return new FixUsingsResponse(docText.ToString(), ambiguous);
+        }
+
+        private static async Task<List<QuickFix>> AddMissingUsings(string fileName, OmnisharpWorkspace workspace, Document document, SemanticModel semanticModel, ICodeActionProvider codeActionProvider)
+        {
+            bool processMore = true;
+            var ambiguousNodes = new List<SimpleNameSyntax>();
+            var ambiguous = new List<QuickFix>();
+
+            while (processMore)
+            {
+                bool updated = false;
+
+                var syntaxNode = (await document.GetSyntaxTreeAsync()).GetRoot();
+                var nodes = syntaxNode.DescendantNodes()
+                    .OfType<SimpleNameSyntax>()
+                    .Where(x => semanticModel.GetSymbolInfo(x).Symbol == null && !ambiguousNodes.Contains(x)).ToList();
+
+                foreach (var node in nodes)
+                {
+                    document = workspace.GetDocument(fileName);
+                    semanticModel = await document.GetSemanticModelAsync();
+                    var sourceText = (await document.GetTextAsync());
+                    var diagnostics = semanticModel.GetDiagnostics();
+                    var location = node.Identifier.Span;
+                    var actions = new List<CodeAction>();
+                    var usingOperations = new Dictionary<string, CodeActionOperation>();
+                    var pointDiagnostics = diagnostics.Where(d => d.Location.SourceSpan.Contains(location)).ToImmutableArray();
```

Comment:
This should be filtered further to only include the relevant diagnostics for this location.